internal/controller/supervisorconfig/oidcclientwatcher/oidc_client_watcher.go

```diff
@@ -138,100 +152,94 @@ func (c *oidcClientWatcherController) Sync(ctx controllerlib.Context) error {
-			Message: fmt.Sprintf("%q must be included in %q when %q is included in %q",
-				offlineAccessScopeName, allowedScopesFieldName, refreshTokenGrantTypeName, allowedGrantTypesFieldName),
-		})
-	case allowedScopesContains(oidcClient, requestAudienceScopeName) &&
-		(!allowedScopesContains(oidcClient, usernameScopeName) || !allowedScopesContains(oidcClient, groupsScopeName)):
-		conditions = append(conditions, &v1alpha1.Condition{
-			Type:   allowedScopesValid,
-			Status: v1alpha1.ConditionFalse,
-			Reason: reasonMissingRequiredValue,
-			Message: fmt.Sprintf("%q and %q must be included in %q when %q is included in %q",
-				usernameScopeName, groupsScopeName, allowedScopesFieldName, requestAudienceScopeName, allowedScopesFieldName),
-		})
-	case allowedGrantTypesContains(oidcClient, tokenExchangeGrantTypeName) && !allowedScopesContains(oidcClient, requestAudienceScopeName):
-		conditions = append(conditions, &v1alpha1.Condition{
-			Type:   allowedScopesValid,
-			Status: v1alpha1.ConditionFalse,
-			Reason: reasonMissingRequiredValue,
-			Message: fmt.Sprintf("%q must be included in %q when %q is included in %q",
-				requestAudienceScopeName, allowedScopesFieldName, tokenExchangeGrantTypeName, allowedGrantTypesFieldName),
-		})
-	default:
+	m := make([]string, 0, 4)
+
+	if !allowedScopesContains(oidcClient, openidScopeName) {
+		m = append(m, fmt.Sprintf("%q must always be included in %q", openidScopeName, allowedScopesFieldName))
+	}
+	if allowedGrantTypesContains(oidcClient, refreshTokenGrantTypeName) && !allowedScopesContains(oidcClient, offlineAccessScopeName) {
+		m = append(m, fmt.Sprintf("%q must be included in %q when %q is included in %q",
+			offlineAccessScopeName, allowedScopesFieldName, refreshTokenGrantTypeName, allowedGrantTypesFieldName))
+	}
+	if allowedScopesContains(oidcClient, requestAudienceScopeName) &&
+		(!allowedScopesContains(oidcClient, usernameScopeName) || !allowedScopesContains(oidcClient, groupsScopeName)) {
+		m = append(m, fmt.Sprintf("%q and %q must be included in %q when %q is included in %q",
+			usernameScopeName, groupsScopeName, allowedScopesFieldName, requestAudienceScopeName, allowedScopesFieldName))
+	}
+	if allowedGrantTypesContains(oidcClient, tokenExchangeGrantTypeName) && !allowedScopesContains(oidcClient, requestAudienceScopeName) {
+		m = append(m, fmt.Sprintf("%q must be included in %q when %q is included in %q",
+			requestAudienceScopeName, allowedScopesFieldName, tokenExchangeGrantTypeName, allowedGrantTypesFieldName))
+	}
+
+	if len(m) == 0 {
 		conditions = append(conditions, &v1alpha1.Condition{
 			Type:    allowedScopesValid,
 			Status:  v1alpha1.ConditionTrue,
 			Reason:  reasonSuccess,
 			Message: fmt.Sprintf("%q is valid", allowedScopesFieldName),
 		})
+	} else {
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:    allowedScopesValid,
+			Status:  v1alpha1.ConditionFalse,
+			Reason:  reasonMissingRequiredValue,
+			Message: strings.Join(m, "; "),
```

Comment:
That a great point. I think they do have a stable ordering in the current implementation, since each potential error case is always evaluated in the same order, and the error messages are always appended to the list before being joined (and join keeps that order), but let me know if you see something that would suggest otherwise.